compiler/fir/resolve/src/org/jetbrains/kotlin/fir/resolve/dfa/cfg/ControlFlowGraphBuilder.kt

```diff
@@ -1348,31 +1348,32 @@ class ControlFlowGraphBuilder {
     // ----------------------------------- Safe calls -----------------------------------
 
     fun enterSafeCall(safeCall: FirSafeCallExpression): EnterSafeCallNode {
-        /*
-         * We create
-         *   lastNode -> enterNode
-         *   lastNode -> exitNode
-         * instead of
-         *   lastNode -> enterNode -> exitNode
-         * because we need to fork flow before `enterNode`, so `exitNode`
-         *   will have unchanged flow from `lastNode`
-         *   which corresponds to a path with nullable receiver.
-         */
-        val lastNode = lastNodes.pop()
         val enterNode = createEnterSafeCallNode(safeCall)
-        lastNodes.push(enterNode)
         val exitNode = createExitSafeCallNode(safeCall)
         exitSafeCallNodes.push(exitNode)
-        addEdge(lastNode, enterNode)
-        if (elvisRhsEnterNodes.topOrNull()?.fir?.lhs === safeCall) {
-            //if this is safe call in lhs of elvis, we make two edges
-            // 1. Df-only edge to exit node, to get not null implications there
-            // 2. Cf-only edge to elvis rhs
+        val lastNode = lastNodes.pop()
+        if (lastNode is ExitSafeCallNode) {
+            // Only the non-null branch of the previous safe call can enter this one.
+            //   a ----> a.b -----> a?.b.c ------> a?.b?.c
+            //       \-----\-> a?.b (null) ---^
+            addEdge(lastNode.lastPreviousNode, enterNode)
+        } else {
+            addEdge(lastNode, enterNode)
+        }
+        val nextElvisRHS = elvisRhsEnterNodes.topOrNull()
+        if (nextElvisRHS?.fir?.lhs === safeCall) {
+            // Can skip the null edge directly to elvis RHS.
+            //                            /-----------v
+            //   a ----> a.b ----> a?.b ----> c ----> a?.b ?: c
+            //       \------------------------^
+            // Only do this for control flow though - while technically data flow is the same,
+            // fewer implications are needed to implement the general version.
```

Comment:
This is actually not true - if both data and control flow are directed to elvis RHS, then the safe call exit node has only one input (non-null receiver case) and thus needs no new implications *at all*. So the total number of implications in the safe call is the same, *and* the liveness of nodes is more correct. I'll do that separately together with fixing other cases where liveness is incorrect.

Reply:
Amended the commit since I had to rebase anyway.